stdlib/domain.ml

```diff
@@ -104,40 +125,75 @@ module DLS = struct
         if idx < s then s else compute_new_size (2 * s)
       in
       let new_sz = compute_new_size sz in
-      let new_st = Array.make new_sz unique_value in
+      let new_st = Array.make new_sz Obj_opt.none in
       Array.blit st 0 new_st 0 sz;
-      set_dls_state new_st;
-      new_st
+      (* We want a implementation that is safe with respect to
+         single-domain multi-threading: retry if the DLS state has
+         changed under our feet.
+         Note that the number of retries will be very small in
+         contended scenarios, as the array only grows, with
+         exponential resizing. *)
+      if compare_and_set_dls_state st new_st
+      then new_st
+      else maybe_grow idx
     end
 
-  let set (idx, _init) x =
+  let set (type a) (idx, _init) (x : a) =
     let st = maybe_grow idx in
     (* [Sys.opaque_identity] ensures that flambda does not look at the type of
      * [x], which may be a [float] and conclude that the [st] is a float array.
      * We do not want OCaml's float array optimisation kicking in here. *)
-    st.(idx) <- Obj.repr (Sys.opaque_identity x)
+    st.(idx) <- Obj_opt.some (Sys.opaque_identity x)
+
 
-  let get (idx, init) =
+  let[@inline never] array_compare_and_set a i oldval newval =
+    (* Note: we cannot use [@poll error] due to the
+       allocations on a.(i) in the Double_array case. *)
+    let curval = a.(i) in
```

Comment:
You could get around this problem by using `Obj.(obj (field (repr a) i))` instead of `a.(i)`.
This amounts to using `Obj.t` as your "uniform array" type.

Reply:
Actually this trick does not work, `Obj.field` gets compiled in the same way as `Array.unsafe_get`, it generates a conditional on the block tag and an allocation in the float path. I could suppress this code path with a well-placed `let a = (Obj.magic a : int array)`, but this feels too evil to me -- I don't know that it is safe.